### PR TITLE
feat(router/atc): extend `route.snis` to support wildcard for the leftmost or rightmost character with traditional-compatible flavor

### PR DESCRIPTION
**Pull request by catbro666**
### Summary

- Change the element type of `route.snis` from sni to wildcard_host.
- Change the router accordingly to support wildcard snis because the sni is involved in route matching.

Note this is to be a hidden feature intentionally so we don't document it.

Additional Notes:

- The expressions flavor doesn't support mTLS (mtls-auth plugin and tls-handshake-modifier plugin) theoretically. The reason is the current mTLS implementation is based on the `route.snis`, but at the ssl phase it is not yet known which route will match. As a workaround, it collects the sni set on routes that are associated with mtls plugins in advance. But for the expressions flavor, things are different. All the fields that were involved in route matching have been merged into the field `expression`. Without actually evaluating, we can't know in general if a sni will match a certain route. But again, you can't get all the parameters required for evaluation at the ssl phase. The correct solution is decoupling the mTLS logic from the `routes` entity by binding it to the `snis` entity, as we’ve discussed in https://konghq.atlassian.net/browse/KAG-3757
- We don't distinguish the priority of plain-only-snis and contain-wildcard-snis for the traditional-compatible flavor because there are no reserved bits available. See https://github.com/Kong/kong/blob/master/kong/router/transform.lua#L515-L528

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

https://konghq.atlassian.net/browse/KAG-3832

### CODE REVIEW

**Comment by chronolaw** [comment]
Should we schedule a short offline meeting to discuss this PR? I am not very clear about the requirement.

Another thing, we are planing a big refactoring of router (#12667), perhaps you should take a look at it.

**Comment by chronolaw** [comment]
Convert to draft because we still have some discussion.

**Comment by chronolaw** [comment]
We should also add compatible check code for hybrid mode, See #12814

**Comment by catbro666** [comment]
> We should also add compatible check code for hybrid mode, See #12814

`typedefs.sni` and `typedefs.wildcard_host` are both of type `string`, just the validation functions are different. And `validate_wildcard_host` is stricter than `validate_sni` (which means those strings can pass `validate_wildcard_host` will certainly pass by `validate_sni`), so I think there is no need to add the compatible check code.

**Comment by chronolaw** [comment]
> > We should also add compatible check code for hybrid mode, See #12814
> 
> `typedefs.sni` and `typedefs.wildcard_host` are both of type `string`, just the validation functions are different. And `validate_wildcard_host` is stricter than `validate_sni` (which means those strings can pass `validate_wildcard_host` will certainly pass `validate_sni`), so I think there is no need to add the compatible check code.

No, atc-router in kong gateway lower than 3.7 does not understand wildcard sni, it only support `sni == a.com`.
If atc-router with flavor `traditional_compatible` get a expression `sni == *.com` it can not work as we wish.

**Comment by catbro666** [comment]
> > > We should also add compatible check code for hybrid mode, See #12814
> > 
> > 
> > `typedefs.sni` and `typedefs.wildcard_host` are both of type `string`, just the validation functions are different. And `validate_wildcard_host` is stricter than `validate_sni` (which means those strings can pass `validate_wildcard_host` will certainly pass `validate_sni`), so I think there is no need to add the compatible check code.
> 
> No, atc-router in kong gateway lower than 3.7 does not understand wildcard sni, it only support `sni == a.com`. If atc-router with flavor `traditional_compatible` get a expression `sni == *.com` it can not work as we wish.

Ah right. `utils.normalize_ip` does this limitation. Will add the compatible check code, Thanks!

File: kong/router/transform.lua
Comment (chronolaw) [inline kong/router/transform.lua:262]: We should not create this new function, instead we should use `gen_for_field` with param `sni_op_transform` and `sni_val_transform`

File: kong/router/transform.lua
Comment (chronolaw) [inline kong/router/transform.lua:348]: Here should be:
```
local gen = gen_for_field("tls.sni", sni_op_transform, snis, sni_val_transform)
```

File: kong/clustering/compat/checkers.lua
Comment (chronolaw) [inline kong/clustering/compat/checkers.lua:35]: I think that snis with wildcard also can not work with traditional route lower than 3.7.
In 3.6 and others the traditional router can not recognize snis like "*.com" or "a.*",
so this logic of flavor is needless.

File: kong/router/transform.lua
Comment (catbro666) [reply kong/router/transform.lua]: applied

File: kong/router/transform.lua
Comment (catbro666) [reply kong/router/transform.lua]: applied

File: kong/router/transform.lua
Comment (chronolaw) [inline kong/router/transform.lua:262]: Could we move `sni_op_transform` and `sni_val_transform` after `is_stream_route`, i.e. the original `sni_val_transform`'s place?
This will be more friendly for reviewers.

File: kong/router/transform.lua
Comment (chronolaw) [inline kong/router/transform.lua:294]: `return sni` at the end of function.

File: kong/router/transform.lua
Comment (catbro666) [reply kong/router/transform.lua:281]: ok

File: kong/router/transform.lua
Comment (catbro666) [reply kong/router/transform.lua]: fixed

File: kong/clustering/compat/checkers.lua
Comment (catbro666) [reply kong/clustering/compat/checkers.lua:35]: Will remove this once your refactor PR is merged. Currently `route.snis` doesn't exist for the `expressions` flavor.

File: kong/router/transform.lua
Comment (chronolaw) [inline kong/router/transform.lua:305]: I think that this block should be the first logic in the function.

File: spec/01-unit/19-hybrid/03-compat_spec.lua
Comment (chronolaw) [inline spec/01-unit/19-hybrid/03-compat_spec.lua:638]: The for loop should include the describe block, like
```
for _, flavor in ipairs({ "traditional", "traditional_compatible"}) do
  describe("Router (flavor = " .. flavor .. ") route entities compatible changes", function()
```

File: kong/router/transform.lua
Comment (catbro666) [reply kong/router/transform.lua:305]: Because the last character can't be `DOT` or `ASTERISK` at the same time, so there is no need to check this when `op == OP_PREFIX`

File: kong/clustering/compat/checkers.lua
Comment (chronolaw) [inline kong/clustering/compat/checkers.lua:42]: Should it be `continue`?

File: spec/01-unit/19-hybrid/03-compat_spec.lua
Comment (catbro666) [reply spec/01-unit/19-hybrid/03-compat_spec.lua]: applied

File: kong/clustering/compat/checkers.lua
Comment (catbro666) [reply kong/clustering/compat/checkers.lua]: yes, fixed

File: kong/clustering/compat/checkers.lua
Comment (catbro666) [reply kong/clustering/compat/checkers.lua]: Indeed.

File: kong/clustering/compat/checkers.lua
Comment (chronolaw) [inline kong/clustering/compat/checkers.lua:40]: This logic only removes the wildcard snis, I am not sure it is the result we expect.
Should we remove the whole route? Could we have more discussion on it?

File: spec/01-unit/19-hybrid/03-compat_spec.lua
Comment (chronolaw) [inline spec/01-unit/19-hybrid/03-compat_spec.lua:636]: Now the compat logic has no relationship with flavor, should we remove this for loop?

File: kong/router/traditional.lua
Comment (dndx) [inline kong/router/traditional.lua:532]: We actually decided to stop adding new features into traditional already and major change on that implementation is now what we would like to see. Any chance this only suports `traditional_compatiable` instead?

File: kong/clustering/compat/checkers.lua
Comment (dndx) [inline kong/clustering/compat/checkers.lua:49]: This may create security vulnerabilities because expected routes no longer match and just fail silently. We should consider not sending the config and output an error instead.

**Review by dndx — CHANGES_REQUESTED** [review]

File: kong/router/traditional.lua
Comment (catbro666) [reply kong/router/traditional.lua:532]: While technically feasible, considering this is more an enhancement than a feature, the inconsistency may bring more bad than good. For the users, they need to be aware of this inconsistency at all times. For the developers, the inconsistency means special treatment here or there, which can increase the cost of code maintenance. I guess the original intent of stopping adding new features into the traditional is to reduce the cost of maintenance

File: kong/clustering/compat/checkers.lua
Comment (VicYP) [reply kong/clustering/compat/checkers.lua:49]: @mheap @veenaraja Please take a look^

Comment (catbro666) [reply kong/clustering/compat/checkers.lua:49]: Michael is also leaning towards the option 1. I will remove the compatibility check code.

File: kong/clustering/compat/checkers.lua
Comment (catbro666) [reply kong/clustering/compat/checkers.lua:40]: Will remove the compatibility check code as per Datong and Michael's suggestion

File: kong/clustering/compat/checkers.lua
Comment (chronolaw) [reply kong/clustering/compat/checkers.lua]: Where is the action "refuse the whole config and directly error out"?

Comment (catbro666) [reply kong/clustering/compat/checkers.lua]: I think DPs will [refuse the config](https://github.com/Kong/kong/blob/master/kong/clustering/data_plane.lua#L254) on [validation failure](https://github.com/Kong/kong/blob/master/kong/db/schema/others/declarative_config.lua#L739) by default and [send the error info](https://github.com/Kong/kong/blob/master/kong/clustering/data_plane.lua#L288) to the CP.

File: kong/router/traditional.lua
Comment (chronolaw) [reply kong/router/traditional.lua:532]: In my opinion, we should not change the code of traditional router.

Comment (catbro666) [reply kong/router/traditional.lua:532]: Okay, since you all think so, I'll remove the traditional router change.

**Comment by chronolaw** [comment]
The code looks good to me, but let's hold this PR until #12677 is merged, because in that PR we changes the schema of route entity, then we must adjust the implement and tests of this PR.

File: changelog/unreleased/kong/route-wildcard-snis.yml
Comment (chronolaw) [inline changelog/unreleased/kong/route-wildcard-snis.yml:1]: We should add a description that this feature is only available when kong runs with the `traditional_compatible` flavor.

Comment (catbro666) [reply changelog/unreleased/kong/route-wildcard-snis.yml]: added

**Comment by chronolaw** [comment]
I have another concern, since `traditional_compatible` should have the same behavior as `traditional`, should we add this feature to the new `expressions` flavor (#12667).
Also see this comment: https://github.com/Kong/kong/pull/12667#issuecomment-2047103237

**Comment by catbro666** [comment]
IMO, if `expressions` flavor is going to be implemented as a superset of `traditional_compatible`, then it is natural to support this in `expressions` as well (actually it will automatically inherit from `traditional_compatible`). And if `traditional_compatible` should have exactly the same behavior as `traditional`, it should also be in `traditional`. It all depends on how we define the relationship between the three.
@chronolaw @dndx What's gonna be the next step, should we add the `traditional` support back?

**Comment by chronolaw** [comment]
I think that we should only change the definition of `snis` in flavor `expressions` to `wildcard_host`, the schema of `traditional` and `traditional_compatible` will have no change.

Let's have a short offline talk after #12667 is merged.

File: spec/01-unit/08-router_spec.lua
Comment (chronolaw) [inline spec/01-unit/08-router_spec.lua:1735]: I think that now this feature also should support 'expressions' flavor,
could you check the test cases to ensure it?
```suggestion
      if flavor == "tradition_compatible" and flavor == "expressions" then
```

File: kong/db/schema/entities/routes.lua
Comment (chronolaw) [inline kong/db/schema/entities/routes.lua:162]: It will apply to 'traditional' flavor, but I think it will ok

File: spec/01-unit/08-router_spec.lua
Comment (catbro666) [reply spec/01-unit/08-router_spec.lua]: Sure. Added "expressions" flavor.

File: kong/db/schema/entities/routes.lua
Comment (catbro666) [reply kong/db/schema/entities/routes.lua]: I exclude the `traditional` flavor now.

File: kong/db/schema/entities/routes.lua
Comment (chronolaw) [inline kong/db/schema/entities/routes.lua:39]: Could we use a simple `kong_router_flavor == "traditional" and ... or ...`?

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1319]: rename to `trad_compat_configs` may be better.

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1336]: should support 'expressions' flavor.

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1462]: same as above.

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1485]: same as above.

File: kong/db/schema/entities/routes.lua
Comment (catbro666) [reply kong/db/schema/entities/routes.lua:39]: I found in some test environments, the `kong.configuration.router_flavor` doesn't have a value so I set the default value `snis_elements_type` to `typedefs.wildcard_host`

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:134]: I think that we need not to define `it_not_trad`, 
just use `if flavor ~= "traditional" then end` around those special cases.

File: spec/01-unit/08-router_spec.lua
Comment (chronolaw) [inline spec/01-unit/08-router_spec.lua:1766]: Do not need this, I think.

File: spec/01-unit/08-router_spec.lua
Comment (chronolaw) [inline spec/01-unit/08-router_spec.lua:1781]: Do not need this, I think.

File: spec/01-unit/08-router_spec.lua
Comment (chronolaw) [inline spec/01-unit/08-router_spec.lua:1795]:
```suggestion
      end -- if flavor ~= "traditional" then
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1432]:
```suggestion
      end  -- if flavor ~= "traditional" then
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1607]:
```suggestion
      end  -- if flavor ~= "traditional" then
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:2044]:
```suggestion
      end  -- if flavor ~= "traditional" then
```

**Comment by chronolaw** [comment]
Should we add back the change log file to tell the user that we have this new feature?

**Comment by catbro666** [comment]
> Should we add back the change log file to tell the user that we have this new feature?

Confirmed with @dndx that this would be a hidden feature so we don't document it.

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1399]:
```suggestion
        it("matches a route based on its leftmost wildcard sni", function()
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1416]:
```suggestion
        it("matches a route based on its rightmost wildcard sni", function()
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1556]:
```suggestion
        it("matches a route based on its leftmost wildcard sni", function()
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:1582]:
```suggestion
        it("matches a route based on its rightmost wildcard sni", function()
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:2005]:
```suggestion
        it("matches a route based on its leftmost wildcard sni", function()
```

File: spec/02-integration/05-proxy/02-router_spec.lua
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:2025]:
```suggestion
        it("matches a route based on its rightmost wildcard sni", function()
```

**Review by catbro666 — COMMENTED** [review]
apply comments

**Review by chronolaw — APPROVED** [review]
Great job!

**Comment by catbro666** [comment]
@chronolaw  Thanks for the efforts in the review!

**Review by jschmid1 — APPROVED** [review]
I have nothing further to add, thanks to the thorough reviews by my peers before me.

**Review by brentos — APPROVED** [review]

**Comment by ms2008** [comment]
@dndx Can you please review this again?

**Review by hanshuebner — APPROVED** [review]

**Review by bungle — APPROVED** [review]

**Pull request merged**
